### PR TITLE
8346017: Socket.connect specified to throw UHE for unresolved address is problematic for SOCKS V5 proxy

### PR DESCRIPTION
**Pull request by vy**
[JDK-8343791](https://bugs.openjdk.org/browse/JDK-8343791) modified `Socket::connect` to throw `UHE` on unresolved addresses – merged in #22160. As reported in [JDK-8346017](https://bugs.openjdk.org/browse/JDK-8346017), this renders `connect()` unusable when there is a (SOCKS) proxy in play, where calling `connect()` using unresolved addresses is a valid operation. This PR

1. Reverts the earlier `Socket::connect` change
2. Updates tests accordingly
3. Enhances tests with proxied socket inputs

These changes require CSR and RN tickets. I will implement them once we agree on a solution here.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8346204](https://bugs.openjdk.org/browse/JDK-8346204) to be approved

### Issues
 * [JDK-8346017](https://bugs.openjdk.org/browse/JDK-8346017): Socket.connect specified to throw UHE for unresolved address is problematic for SOCKS V5 proxy (**Bug** - P2)(⚠️ The fixVersion in this issue is [24] but the fixVersion in .jcheck/conf is 25, a new backport will be created when this pr is integrated.)
 * [JDK-8346204](https://bugs.openjdk.org/browse/JDK-8346204): Socket.connect specified to throw UHE for unresolved address is problematic for SOCKS V5 proxy (**CSR**)

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22719/head:pull/22719` \
`$ git checkout pull/22719`

Update a local copy of the PR: \
`$ git checkout pull/22719` \
`$ git pull https://git.openjdk.org/jdk.git pull/22719/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22719`

View PR using the GUI difftool: \
`$ git pr show -t 22719`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22719.diff">https://git.openjdk.org/jdk/pull/22719.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22719#issuecomment-2539737067)
</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
The method description of both Socket.connect methods will require adjustment to remove "If the endpoint is an unresolved  InetSocketAddress". I think better to leave the @throws UHE but change the description to "If the connection could not be established because the socket address is unresolved".

**Comment by vy** [comment]
> The method description of both Socket.connect methods will require adjustment to remove "If the endpoint is an unresolved InetSocketAddress". I think better to leave the @throws UHE but change the description to "If the connection could not be established because the socket address is unresolved".

@AlanBateman, implemented this change in ca0bef4044e5a0cba7e972be6277292f2d774b37.

**Comment by AlanBateman** [comment]
The update to Socket looks okay, thanks.

ConnectFailTest is a test for connect failing. I'm tempted we create a new test for SOCKS and not modify ConnectFailTest. With SOCK then the connect may succeed with an unresolved address so I think different to the direct case. Daniel, do you agree on this?

**Comment by dfuch** [comment]
> ConnectFailTest is a test for connect failing. I'm tempted to suggest that we create a new test for SOCKS and not modify ConnectFailTest. With SOCKS then the connect may succeed with an unresolved address so I think different to the direct case. Daniel, do you agree on this?

It's a good suggestion. Maybe we could had a test for the HTTP proxy case as well?

**Comment by vy** [comment]
@AlanBateman, @dfuch, implemented the following two changes:

1. Added `new Socket(Proxy.NO_PROXY)` case to `ConnectFailTest`
2. Added `ConnectSocksProxyTest` cloning `ConnectFailTest` for a SOCKS proxy socket

Together with @dfuch, we decided to postpone the HTTP proxy socket case to a separate future PR, since there is more work there needs to be done there.

File: test/jdk/java/net/Socket/ConnectFailTest.java [hunk 1]
```diff
@@ -161,7 +162,8 @@ static List<Socket> sockets() throws Exception {
         Socket socket = new Socket();
         @SuppressWarnings("resource")
         Socket channelSocket = SocketChannel.open().socket();
-        return List.of(socket, channelSocket);
+        Socket proxiedSocket = new Socket(Proxy.NO_PROXY);
```
Comment (AlanBateman) [inline test/jdk/java/net/Socket/ConnectFailTest.java:165]: Adding a Socket created with NO_PROXY is a good idea. On naming "proxiedSocket" suggests it's proxies, I think you want noProxySocket :-)

File: test/jdk/java/net/Socket/ConnectSocksProxyTest.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+            @Override
+            protected PasswordAuthentication getPasswordAuthentication() {
+                return new PasswordAuthentication(PROXY_AUTH_USERNAME, PROXY_AUTH_PASSWORD.toCharArray());
+            }
+        });
+    }
+
+    @BeforeAll
+    static void initProxyServer() throws IOException {
+        PROXY_SERVER = new SocksServer(0);
```
Comment (dfuch) [inline test/jdk/java/net/Socket/ConnectSocksProxyTest.java:82]: We should use the constructor that takes an InetAddress and pass the loopback address here.

File: test/jdk/java/net/Socket/ConnectSocksProxyTest.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        }
+    }
+
+    /**
+     * Delegates to {@link #testUnconnectedSocketWithUnresolvedAddress(boolean, Socket)} using a bound socket.
+     */
+    @Test
+    void testBoundSocketWithUnresolvedAddress() throws IOException {
+        try (Socket socket = createProxiedSocket()) {
+            socket.bind(new InetSocketAddress(0));
```
Comment (dfuch) [inline test/jdk/java/net/Socket/ConnectSocksProxyTest.java:164]: if the server is bound to the loopback, we should use the loopback here too

File: test/jdk/java/net/Socket/ConnectSocksProxyTest.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+    /**
+     * Verifies the behaviour of an unconnected socket when {@code connect()} is invoked using an unresolved address.
+     */
+    private static void testUnconnectedSocketWithUnresolvedAddress(boolean bound, Socket socket) throws IOException {
+        assertEquals(bound, socket.isBound());
+        assertFalse(socket.isConnected());
+        try (ServerSocket serverSocket = createEphemeralServerSocket()) {
+            InetSocketAddress unresolvedAddress =
+                    InetSocketAddress.createUnresolved("localhost", serverSocket.getLocalPort());
```
Comment (dfuch) [inline test/jdk/java/net/Socket/ConnectSocksProxyTest.java:177]:
```suggestion
                    InetSocketAddress.createUnresolved(getLoopbackAddress().getHostAddress(), serverSocket.getLocalPort());
```

**Comment by dfuch** [comment]
Sorry for my comments on outdated code - my review had been pending and I'd forgotten to commit it. I've made the comments again on the new code.

File: test/jdk/java/net/Socket/ConnectSocksProxyTest.java (hunk 4 above)
Comment (vy) [reply test/jdk/java/net/Socket/ConnectSocksProxyTest.java]: Fixed in 9706ef9c7c362829daa7cc3ea845a85e36bf92c4.

File: test/jdk/java/net/Socket/ConnectSocksProxyTest.java (hunk 2 above)
Comment (vy) [reply test/jdk/java/net/Socket/ConnectSocksProxyTest.java]: Fixed in 9706ef9c7c362829daa7cc3ea845a85e36bf92c4.

File: test/jdk/java/net/Socket/ConnectFailTest.java [hunk 5]
```diff
@@ -81,8 +113,8 @@ void testUnboundSocket(Socket socket) throws IOException {
      */
     @ParameterizedTest
     @MethodSource("sockets")
-    void testBoundSocket(Socket socket) throws IOException {
-        try (socket) {
+    void testBoundSocket(SocketArg socketArg) throws IOException {
+        try (Socket socket = socketArg.socket) {
             socket.bind(new InetSocketAddress(0));
```
[the remark this answers is not in the document]
Comment (vy) [reply test/jdk/java/net/Socket/ConnectFailTest.java]: Fixed in 9706ef9c7c362829daa7cc3ea845a85e36bf92c4.

File: test/jdk/java/net/Socket/ConnectFailTest.java [hunk 6]
```diff
@@ -112,29 +145,49 @@ void testConnectedSocket(Socket socket) throws Throwable {
...
-     * Verifies that a bound socket is closed when {@code connect()} is invoked using an unresolved address.
+     * Delegates to {@link #testUnconnectedSocketWithUnresolvedAddress(boolean, SocketArg)} using a bound socket.
      */
     @ParameterizedTest
     @MethodSource("sockets")
-    void testBoundSocketWithUnresolvedAddress(Socket socket) throws IOException {
-        try (socket) {
+    void testBoundSocketWithUnresolvedAddress(SocketArg socketArg) throws IOException {
+        try (Socket socket = socketArg.socket) {
             socket.bind(new InetSocketAddress(0));
```
[the remark this answers is not in the document]
Comment (vy) [reply test/jdk/java/net/Socket/ConnectFailTest.java]: Fixed in 9706ef9c7c362829daa7cc3ea845a85e36bf92c4.

File: test/jdk/java/net/Socket/ConnectFailTest.java [hunk 7]
```diff
@@ -57,18 +62,45 @@ class ConnectFailTest {
...
+            @Override
+            protected PasswordAuthentication getPasswordAuthentication() {
+                return new PasswordAuthentication(SOCKS_AUTH_USERNAME, SOCKS_AUTH_PASSWORD.toCharArray());
+            }
+        });
+    }
+
+    @BeforeAll
+    static void initSocksServer() throws IOException {
+        SOCKS_SERVER = new SocksServer(0);
```
[the remark this answers is not in the document]
Comment (vy) [reply test/jdk/java/net/Socket/ConnectFailTest.java]: Fixed in 9706ef9c7c362829daa7cc3ea845a85e36bf92c4.

**Comment by vy** [comment]
Added successful `tier1,2` test reports generated using 9706ef9c7c362829daa7cc3ea845a85e36bf92c4 to the ticket.

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**